### ISSUE TITLE
Infer closure param types from return type closure params

### ISSUE THREAD

**Issue by muglug** [vimeo/psalm#2896]
These three should not have any errors:

```php
/**
 * @template T
 * @template U
 * @param callable(T): U $predicate
 * @return Closure(iterable<T>): iterable<U>
 */
function map(callable $predicate): callable {
    return function($iter) use ($predicate): iterable {
        foreach ($iter as $key => $value) {
            yield $key => $predicate($value);
        }
    };
}
```

```php
/**
 * @template T
 * @template U
 * @param callable(T): U $predicate
 * @return Closure(iterable<T>): iterable<U>
 */
function map(callable $predicate): callable {
    return function(iterable $iter) use ($predicate): iterable {
        foreach ($iter as $key => $value) {
            yield $key => $predicate($value);
        }
    };
}
```

```php
/**
 * @template T
 * @template U
 * @param callable(T): U $predicate
 * @return callable(iterable<T>): iterable<U>
 */
function map(callable $predicate): callable {
    return function($iter) use ($predicate): iterable {
        foreach ($iter as $key => $value) {
            yield $key => $predicate($value);
        }
    };
}
```

But this should throw an error:

```php
/**
 * @template T
 * @template U
 * @param callable(T): U $predicate
 * @return callable(iterable<T>): iterable<U>
 */
function map(callable $predicate): callable {
    $a = function($iter) use ($predicate): iterable {
        foreach ($iter as $key => $value) {
            yield $key => $predicate($value);
        }
    };
    return $a;
}
```

**Comment by ragboyjr** [comment]
Spent a bit more time on this and found a few other examples that we should probably support in this:

```php
// inferring return type from closure typings.
/**
 * @template T
 * @template U
 * @param callable(T): U $predicate
 */
function map(callable $predicate): callable {
    return 
    /**
     * @param iterable<T>
     * @return iterable<U>
     */
    function($iter) use ($predicate): iterable {
        foreach ($iter as $key => $value) {
            yield $key => $predicate($value);
        }
    };
}
```

This is the valid typescript equivalent:

```typescript
function map<T, U>(fn: (t: T) => U) {
  return (iter: T[]): U[] => {
    let newIter: U[] = [];
    iter.forEach(v => newIter.push(fn(v)))
    return newIter;
  }
}
```

And then finally, (i'm not sure if this should be a new ticket on its own), but allowing generic anonymous functions:

Typescript syntax:

```typescript
function createGeneric() {
  // the extends unknown is just a hack to get around parsing limitations
  return <T extends unknown, U extends unknown>(fn: (t: T) => U, v: T): U => {
    return fn(v);
  }
}
```

```php
function createGeneric() {
  return
  /**
   * @template T
   * @template U
   * @param callable(T): U $fn
   * @param T $v
   * @return U
   */
   function(callable $fn, $v) { return $fn($v); };
}
```

The only reason i'm including this is because I don't believe there's anyway to support this OTHER than type inference.
